### ISSUE TITLE
question on sparseDecom params

### ISSUE THREAD

**Issue by portokalh** [ANTsX/ANTsR#229]
I am trying to get a sense for what are good ell1 values (i have it set it to 1), to improve speed & convergence

i do get an occasional error :
Error: vector memory exhausted (limit reached?)

and bonus warnings e.g.:
In cor(inmatrix[, i], reconmat[, i]) : the standard deviation is zero

What i think is that  at each iteration i get for each eigenvector in the set a Rayleigh quotient (maybe V*Av), and a norm (fnz?) for the error norm(Ax-lambda x). But i do not see that changing much, so i am thinking i may not have gotten things interpreted right. 

Any thoughts? 

Thanks much!

alex

**Comment by muschellij2** [comment]
Have you checked your memory usage while using this?

John

**Comment by portokalh** [comment]
I have 64GB memory, using <2.6 , and use a single CPU so there should be plenty of room; and I can in fact run multiple processes in parallel, until the last step.

The error occurs at the last iteration, no matter how many iterations i set. If i set 1 it happens after 1; if i set 5 it happens at the 5th iteration. I have switched to powerit. 

eanat_mang <- sparseDecom( inmatrix=mang.train, inmask=mask, nvecs=mynvecs,
                             sparseness=mysp, cthresh=250, its=1,  mycoption=0, verbose=1, powerit=1, statdir=paste0(output.path,'/sp5_20vec/')) 

[1] 1
 sparse recon 250 useL1 1
rayleigh-quotient: 4.24298e+13 in 1 num 0
rayleigh-quotient: 2.49794e+13 in 1 num 0
rayleigh-quotient: 1.38147e+13 in 1 num 0
rayleigh-quotient: 1.59565e+13 in 1 num 0
rayleigh-quotient: 2.20295e+13 in 1 num 0
rayleigh-quotient: 1.21835e+13 in 1 num 0
rayleigh-quotient: 1.59366e+13 in 1 num 0
rayleigh-quotient: 8.30843e+12 in 1 num 0
rayleigh-quotient: 2.03814e+13 in 1 num 0
rayleigh-quotient: 1.45935e+13 in 1 num 0
rayleigh-quotient: 1.60182e+13 in 1 num 0
rayleigh-quotient: 1.33683e+13 in 1 num 0
rayleigh-quotient: 1.60776e+13 in 1 num 0
rayleigh-quotient: 1.71126e+13 in 1 num 0
rayleigh-quotient: 4.86866e+13 in 5 num 0
rayleigh-quotient: 4.86534e+13 in 6 num 0
rayleigh-quotient: 4.86539e+13 in 6 num 0
rayleigh-quotient: 4.86539e+13 in 6 num 0
rayleigh-quotient: 4.86539e+13 in 6 num 0
rayleigh-quotient: 8.67224e+12 in 1 num 0
0: %var 0.899545 raw-reconerr 17318.1
Error: vector memory exhausted (limit reached?)
In addition: There were 50 or more warnings (use warnings() to see the first 50)
> warnings()
Warning messages:
1: In cor(inmatrix[, i], reconmat[, i]) : the standard deviation is zero
I find this warning worrisome.

**Comment by portokalh** [comment]
i hope it is not R 3.5
https://stat.ethz.ch/pipermail/r-help//2018-June/454855.html

**Comment by dorianps** [comment]
They show a solution a few mails forward, i.e., R_MAX_VSIZE=700Gb is woking
on 3.5 Mac using RStudio.

Is that doing the trick for you? Are you indeed on a Mac as the person
reporting the issue above?

**Comment by portokalh** [comment]
I am using MAC OSX 10.11.6 
i have tried and it does not completely fix the problem. 
I have added 

gc(verbose = TRUE, reset = FALSE, full = TRUE)

then tried

 jeanat_mang <- joinEigenanatomy( mang.train, mask, eanat_mang$eigenanatomyimages, graphdensity=0.1, joinMethod='multilevel' ,verbose=TRUE)

and i get again:
Error: vector memory exhausted (limit reached?)

**Comment by muschellij2** [comment]
Can you provide a MCWE: https://stackoverflow.com/help/mcve?  Otherwise, these issues are almost impossible to try and resolve.

Did you Restart R, run:

```r
Sys.setenv("R_MAX_VSIZE" = 8e9) 
```

And retry the command?

**Comment by dorianps** [comment]
Note that the variable is read at startup, so changing it with Sys* from R
itself doesn't have any effect. The temporary solution that worked for the
guy was to include "R_MAX_VSIZE=700Gb" in .Renviron.

**Comment by portokalh** [comment]
It seems strange, but it could be that the last error is due to me getting the mask format wrong, since the following line works? 
My mask is an antsImage. 
 
jeanat_mang <- joinEigenanatomy( mang.train, mask=NA, eanat_mang$eigenanatomyimages, graphdensity=0.1, joinMethod='multilevel', verbose=TRUE)

I need to apologize at this point since this was not an error message related to memory, but it has not been straight fwd for me to think of the mask, when reading

Error: vector memory exhausted (limit reached?)

I hope to not see the memory error message again too soon. 

Thank you all! 
Alex

**Comment by stnava** [comment]
i am not sure which issue needs to be addressed here but recommend following the examples given for `sparseDecom` that use images.   for example:  

[https://github.com/stnava/blindSourceSeparationInANTsR](https://github.com/stnava/blindSourceSeparationInANTsR/)

**Comment by portokalh** [comment]
I was wondering about sparseDecom, and these questions still stand (for me):

1. recommended step sizes for ell1
2. the proper way to follow how it converges

For the rest, a short summary is:
The memory related errors were apparently not memory related after all.  
I have ended up not using a mask in JoinEigenenatomy , which must be be my bad, but i have tried three formats and things did not go well. 
Besides this I have updated RStudio. 

And it seems that things go well after that, thanks to all for illuminating the way!
